core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
+    this map can be serialized and sent to OMS as is */
+    private ConcurrentHashMap<InetSocketAddress, NodeMetric> metrics;
 
     private static final Logger logger = Logger.getLogger(KernelClient.class.getName());
 
+    public ConcurrentHashMap<InetSocketAddress, NodeMetric> getMetrics() {
+        return metrics;
+    }
+
     /**
-     * Add a host to the list of hosts that we've contacted
+     * Adds remote kernel server information to cache for the given host
      *
      * @param host
      */
-    private KernelServer addHost(InetSocketAddress host) {
+    private KernelServerInfo addHost(InetSocketAddress host) {
         try {
             Registry registry = LocateRegistry.getRegistry(host.getHostName(), host.getPort());
             KernelServer server = (KernelServer) registry.lookup("io.amino.run.kernelserver");
-            servers.put(host, server);
-            return server;
+            NodeMetric metric = new NodeMetric();
+            KernelServerInfo serverInfo = new KernelServerInfo(host, server, metric);
+            servers.put(host, serverInfo);
+            metrics.put(host, metric);
+            return serverInfo;
         } catch (Exception e) {
-            logger.severe("Could not find MicroService server on host: " + e.toString());
+            logger.severe(
+                    String.format(
+                            "Could not find kernel server on host: %s. Exception: %s", host, e));
         }
         return null;
     }
 
+    /**
+     * Removes cached remote kernel server information for the given host, if it has missed
+     * heartbeats for {@link amino.run.kernel.client.KernelClient#MAX_HEARTBEAT_MISS_TIMES} times
+     *
+     * @param host
+     */
+    private void removeHost(InetSocketAddress host) {
+        KernelServerInfo server = servers.get(host);
+        if (server != null) {
+            if ((server.refCount - server.refCountOnHeartBeat >= MAX_HEARTBEAT_MISS_TIMES)) {
+                servers.remove(host);
+                metrics.remove(host);
+            }
+        }
+    }
+
+    /**
+     * Gets the remote kernel server information for the given host
+     *
+     * @param host
+     * @return Kernel Server Information
+     */
+    private KernelServerInfo getServerInfo(InetSocketAddress host) {
+        KernelServerInfo serverInfo = servers.get(host);
+        if (serverInfo == null) {
+            serverInfo = addHost(host);
+        }
+        return serverInfo;
+    }
+
+    /**
+     * Gets the RMI stub reference to remote kernel server for the given host
+     *
+     * @param host
+     * @return RMI stub to remote kernel server
+     */
     private KernelServer getServer(InetSocketAddress host) {
-        KernelServer server = servers.get(host);
-        if (server == null) {
-            server = addHost(host);
+        KernelServerInfo serverInfo = getServerInfo(host);
+        return serverInfo != null ? serverInfo.remoteRef : null;
```

Comment:
I don't like passing nulls around like this.  People forget to check return values for null and then we get intermittent memory problems, rather than easier-to-debug exceptions.  Rather throw an exception?

Reply:
PS: Because getServerInfo() never returns null, this method should also never need to return null.  If getServerInfo() returns null then we have a problem and throwing an exception is appropriate.

Reply:
Agreed. Throwing `KernelServerNotFoundException` exception from `addHost()` upon failure and propagating it to `getServerInfo()` and `getServer()`.